crates/uv/src/commands/python/pin.rs

```diff
@@ -82,3 +122,102 @@ pub(crate) async fn pin(
 
     Ok(ExitStatus::Success)
 }
+
+/// Check if pinned request is compatible with the workspace/project's `Requires-Python`.
+fn check_request_requires_python_compatibility(
+    pin: &PythonRequest,
+    virtual_project: &VirtualProject,
+    python_preference: PythonPreference,
+    cache: &Cache,
+) {
+    let requested_version = match pin {
+        PythonRequest::Version(ref version) => {
+            let version = pep440_rs::Version::from_str(&version.to_string());
+            match version {
+                Ok(version) => Some(version),
+                Err(e) => {
+                    debug!("Failed to parse PEP440 python version from {pin}: {e}");
+                    None
+                }
+            }
+        }
+        _ => None,
+    };
+
+    // Check if the requested version is compatible with the project.
+    // If the compatibility check fails, exit early.
+    if let Some(version) = requested_version {
+        if let Err(e) = assert_python_compatibility(&version, virtual_project) {
+            warn_user_once!("{}", e);
+            return;
+        }
+    };
+
+    // If the requested version is either not specified or compatible, attempt to resolve the request into an interpreter.
+    let python_version = match PythonInstallation::find(
+        pin,
+        EnvironmentPreference::OnlySystem,
+        python_preference,
+        cache,
+    ) {
+        Ok(python) => Ok(python.python_version().clone()),
+        Err(err) => Err(err.to_string()),
+    };
+
+    match python_version {
+        Ok(python_version) => {
+            debug!(
+                "The pinned Python version {} resolves to {}",
+                pin, python_version
+            );
+            if let Err(e) = assert_python_compatibility(&python_version, virtual_project) {
+                warn_user_once!("{}", e);
+            }
+        }
+        Err(e) => {
+            warn_user_once!(
+                "Failed to resolve pinned Python version from {}: {}",
```

Comment:
We should use `PythonRequest::to_canonical_string` for display here and wrap in backticks.